compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOverrideChecker.kt

```diff
@@ -130,27 +130,35 @@ object FirOverrideChecker : FirRegularClassChecker() {
         }
     }
 
+    // See [OverrideResolver#isReturnTypeOkForOverride]
     private fun FirCallableMemberDeclaration<*>.checkReturnType(
         overriddenSymbols: List<FirCallableSymbol<*>>,
         typeCheckerContext: AbstractTypeCheckerContext,
         context: CheckerContext,
     ): FirMemberDeclaration? {
-        val returnType = returnTypeRef.safeAs<FirResolvedTypeRef>()?.type
+        val overridingReturnType = returnTypeRef.safeAs<FirResolvedTypeRef>()?.type
```

Comment:
Minor: use `FirTypeRef.coneType` instead of manual casting. Also safe casting is redundant on checkers stage because it's guaranteed that all type refs are resolved (and if it's not then it's a bug which we should find AFAP)

Reply:
It was just renaming to be clearer, but thanks for the insight as well.

I do remember seeing many other cases in checker, and thus attempted to remove all of such redundant casting to resolved type ref in checkers. See 5th commit.

Alas, not all are converted at this moment. It looks like, when the input code is broken/incomplete, e.g., incomplete try/catch, incomplete class definition, there could be still implicit type ref in catch parameter or value parameter. I left TODOs in checkers, but most likely the work should be done during resolution I guess.

I have tried to convert remaining named reference without candidate into an unresolved reference (as an attempt for https://youtrack.jetbrains.com/issue/KT-43948), but was not able to make it, i.e., can't find a right place to do so. I'm a bit worried that this---remaining implicit type ref---would be similar as well, but maybe not. Any suggestions, e.g., places to look at, would be helpful. E.g., for catch parameter, `FirControlFlowStatementsResolveTransformer#transformCatch` might be the place where we can replace implicit return type of catch, if it's still implicit after `transformParameter`. Not sure about the implicit type of value parameter.

Anyway, those could be done saparately.

Reply:
Ideally all named references should be converted to resolved/error reference on (implicit) body resolve stage, so all fixes of it should be somewhere in body resolve transformers

Types are converted in three phases, so there are three phases for fixing implicits:
- supertype resolve transformer
- type resolve transformer (for types of members, value parameters)
- body resolve transformers (for every type in local scope of functions/property body)

Reply:
Thanks for the pointers. I think 2nd/3rd resolve transformers would be right places for implicit types I've seen.

My question is, would you want me to keep stacking commits on this PR or can it be done separately? I heard that extending override checker to implicit return type (as part of #4060) revealed a hidden property type mismatch, which turns out to be a false alarm, since the last safe call of a lambda wasn't retrieved as the lambda return. Fixing that also revealed an issue of coercion-to-Unit for early returned lambda. I think up to 4th commit is the right checking point for this PR. If you don't mind, 5th commit is okay to go I think. Then, I can keep digging remaining implicit type ref at checkers stage.

Reply:
It's better to create new PR. I already merged this one

Reply:
Great, thanks. Will keep this thread open to not forget.

Reply:
TODOs I left will be gone here: https://github.com/JetBrains/kotlin/pull/4102